### PR TITLE
SelectAsync

### PR DESCRIPTION
**Pull request by andrewjtait**
SelectAsync renders a regular Select, but wraps it in additional functionality to allow it to fetch its options from an API endpoint.
 
### The Request

By default the component will execute a request with the following options:

```
{
  params: {
    page: 1
    items_per_page: 20,
    search: 'any filter text'
  }
}
```

This can be customised using the formatRequest prop, for example to enable withCredentials:

```
<SelectAsync
  endpoint='/endpoint'
  formatRequest={ opts => ({ ...opts, withCredentials: true }) }
/>
```

The opts object is used with axios, please see this URL for more options: https://github.com/axios/axios#request-config

### The Response
By default the response is expected in the following format:

```
{
  $items: [Array] - required, the items to render
  $page: [Integer] - optional, for paginated responses
  $total: [Integer] - optional, for paginated responses
}
```

If your response does not match this format you can modify it using the formatResponse prop:

```
<SelectAsync
  endpoint='/endpoint'
  formatResponse={ response => ({ ...response, data: { $items: response.myItems } }) 
/>
```

This follows the axios response schema: https://github.com/axios/axios#response-schema

### Rendering Options

By default options will be rendered under the assumption they have the following format:
 
```
{
  id [Integer] - the unique identifier
  displayed_as [String] - the human readable value
}
```

If your items do not match this format, or you want to have custom control over how they are rendered, you can provide a function as a child:
 
```
<SelectAsync endpoint='/endpoint'>
  {
    items => items.map(item => <Option>{ item.text }</Option>)
  }
</SelectAsync>
```

### CODE REVIEW

File: src/__experimental__/components/select-async/select-async.component.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+      this.props.endpoint,
+      this.fetchOptions(page, search)
+    );
+    this.handleResponse(data);
+  }
+
+  fetchNextPage = () => {
+    if (this.state.fetching) return;
+    if (this.state.items.length === this.state.total) return;
+    this.fetchData({ page: this.state.page + 1 });
```
Comment (grabkowski) [inline src/__experimental__/components/select-async/select-async.component.js]: object destructuring could be used here, to refer ```this.state``` only once

File: src/__experimental__/components/select-async/select-async.component.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  buildOptions(page, items) {
+    return (page > 1) ? this.state.items.concat(items) : items;
+  }
+
+  renderOptions() {
+    if (this.props.children) return this.props.children(this.state.items);
+
+    if (!this.state.items.length) return null;
+
+    return this.state.items.map(item => (
```
Comment (grabkowski) [inline src/__experimental__/components/select-async/select-async.component.js]: ```this.state.items``` could be a const

File: src/components/scrollable-list/__spec__.js [hunk 3]
```diff
@@ -320,13 +320,13 @@ describe('ScrollableList', () => {
...
+      scrollableList.simulate('scroll', { target: { scrollHeight: 400, scrollTop: 101 } });
       expect(onLazyLoad).toBeCalled();
     });
 
     it('does not trigger onLazyLoad when the scrollTop is not beyond the scrollHeight', () => {
       onLazyLoad.mockReset();
-      scrollableList.simulate('scroll', { target: { scrollHeight: 300, scrollTop: 99 } });
+      scrollableList.simulate('scroll', { target: { scrollHeight: 400, scrollTop: 99 } });
       expect(onLazyLoad).not.toBeCalled();
     });
```
Comment (grabkowski) [inline src/components/scrollable-list/__spec__.js]: shouldn't that file be renamed to: ```scrollable-list.spec.js```?

File: src/__experimental__/components/select-async/select-async.component.js (hunk 1 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: ✅

File: src/__experimental__/components/select-async/select-async.component.js (hunk 2 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: ✅

File: src/components/scrollable-list/__spec__.js (hunk 3 above)
Comment (andrewjtait) [reply src/components/scrollable-list/__spec__.js]: ✅

File: src/__experimental__/components/select-async/select-async.spec.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      const wrapper = await openResults(render());
+      await getNextPage(wrapper);
+      expect(findResults(wrapper)).toMatchSnapshot();
+    });
+
+    it('does not refetch if a fetch is already in progress', async () => {
+      mockResponse();
+      const wrapper = await openResults(render());
+      expect(axios.get.mock.calls.length).toEqual(1);
+      getNextPage(wrapper);
```
Comment (samhinton88) [inline src/__experimental__/components/select-async/select-async.spec.js]: Missing an `await` here I think.

File: src/__experimental__/components/select-async/select-async.spec.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    });
+  });
+
+  describe('custom response', () => {
+    it('throws an error if the response is not supported', async () => {
+      mockResponse({ data: {} });
+      try {
+        await render().instance().fetchData({ page: 1 });
+      } catch ({ message }) {
+        expect(message).toEqual(`The server response does not contain an $items attribute.
```
Comment (samhinton88) [inline src/__experimental__/components/select-async/select-async.spec.js]: Perhaps pop this in a different file and export it to both the test and component, so a change to error message doesn't break the test.

File: src/__experimental__/components/select-async/select-async.spec.js (hunk 4 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.spec.js]: nah this is the subtle difference between this test and the next, this one is testing if a fetch is already in progress (so i do not want to wait for it to finish) maybe i should write a comment next to this line to that effect

File: src/__experimental__/components/select-async/select-async.spec.js (hunk 5 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.spec.js]: alrighty

File: src/__experimental__/components/select-async/select-async.component.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      }
+    };
+
+    if (this.props.formatRequest) opts = this.props.formatRequest(opts);
+
+    return opts;
+  }
+
+  handleResponse(responseData) {
+    let data = responseData;
```
[the remark this answers is not in the document]
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: The problem was a I cannot re-assign `responseData` - do you know if I can write a different way around that?

File: src/__experimental__/components/select-async/select-async.component.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+  filterValue() {
+    return this.select.current.state.filter;
+  }
+
+  fetchOptions(page, search) {
+    let opts = {
+      params: {
+        page,
+        items_per_page: 20,
```
[the remark this answers is not in the document]
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: I can pass this in as a prop 👍

[the remark this answers is not in the document]
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: ✅

File: src/__experimental__/components/select-async/select-async.component.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.component.js]: ✅ changed it a little bit

File: src/__experimental__/components/select-async/select-async.spec.js (hunk 5 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.spec.js]: ✅

File: src/__experimental__/components/select-async/select-async.spec.js (hunk 4 above)
Comment (andrewjtait) [reply src/__experimental__/components/select-async/select-async.spec.js]: ✅

**Pull request merged**